File: src/main/java/com/exampledigisphere/vitapetcare/admin/user/resource/UserResource.java
package com.exampledigisphere.vitapetcare.admin.user.resource;

import com.exampledigisphere.vitapetcare.admin.user.UserDTO;
import com.exampledigisphere.vitapetcare.admin.user.UserService;
import com.exampledigisphere.vitapetcare.admin.user.domain.UserAssociations;
import com.exampledigisphere.vitapetcare.config.root.Info;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.Collections;
import java.util.Set;

@RestController
@PreAuthorize("hasRole('ROLE_ADMIN')")
@RequestMapping("/api/users")
@Info(
  dev = Info.Dev.heltonOliveira,
  label = Info.Label.feature,
  date = "06/02/2026",
  description = "Recurso REST para gestão de usuários"
)
public class UserResource {

  private final UserService userService;

  public UserResource(final UserService userService) {
    this.userService = userService;
  }

  @PostMapping
  @PreAuthorize("hasAuthority('USER_CREATE')")
  @Transactional
  public ResponseEntity<UserDTO> create(@RequestBody @Valid UserDTO input, UriComponentsBuilder uriBuilder) {
    return userService.register(input, Collections.emptySet())
      .map(usr -> {
        var uri = uriBuilder.path("/api/users/{id}").buildAndExpand(usr.getId()).toUri();
        return ResponseEntity.created(uri).body(usr);
      })
      .orElse(ResponseEntity.status(HttpStatus.CONFLICT).build());
  }

  @PutMapping
  @PreAuthorize("hasAuthority('USER_EDIT')")
  @Transactional
  public ResponseEntity<UserDTO> update(@RequestBody @Valid UserDTO input) {
    return userService.register(input, Collections.emptySet())
      .map(ResponseEntity::ok)
      .orElse(ResponseEntity.status(HttpStatus.CONFLICT).build());
  }

  @GetMapping("/{id}")
  @PreAuthorize("hasAuthority('USER_VIEW')")
  public ResponseEntity<UserDTO> getOne(@PathVariable Long id) {
    return userService.retrieve(id, Set.of(UserAssociations.FILES, UserAssociations.WORK_DAYS))
      .map(ResponseEntity::ok)
      .orElse(ResponseEntity.notFound().build());
  }

  @GetMapping
  @PreAuthorize("hasAuthority('USER_VIEW_LIST')")
  public ResponseEntity<Page<UserDTO>> getAll(@PageableDefault(page = 0, size = 20) Pageable pageable) {
    return ResponseEntity.ok(userService.list(pageable));
  }

  @DeleteMapping("/{id}")
  @PreAuthorize("hasAuthority('USER_DELETE')")
  public ResponseEntity<?> disable(@PathVariable Long id) {
    userService.suspend(id);
    return ResponseEntity.ok().build();
  }
}

File: src/main/kotlin/com/exampledigisphere/vitapetcare/user/resource/UserResource.kt
package com.exampledigisphere.vitapetcare.user.resource

import com.exampledigisphere.vitapetcare.user.domain.User
import com.exampledigisphere.vitapetcare.user.useCases.CreateUser
import com.exampledigisphere.vitapetcare.user.useCases.GetAll
import com.exampledigisphere.vitapetcare.user.useCases.GetUser
import jakarta.validation.ConstraintViolationException
import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.util.UriComponentsBuilder

@RestController
@RequestMapping("/api/users")
class UserResource(
  private val createUser: CreateUser,
  private val getUser: GetUser,
  private val getAll: GetAll,
) {

  @PostMapping
  fun create(@RequestBody @Valid input: User, uriBuilder: UriComponentsBuilder): ResponseEntity<*> =
    createUser.execute(input)
      .fold(
        onFailure = { error ->
          error.printStackTrace()
          val status = when (error) {
            is ConstraintViolationException -> HttpStatus.CONFLICT
            else -> HttpStatus.INTERNAL_SERVER_ERROR
          }
          ResponseEntity.status(status).body(null)
        },
        onSuccess = { usr ->
          ResponseEntity
            .created(uriBuilder.path("/api/users/${usr.id}").build().toUri())
            .body(usr.also { it.loadRole() })
        }
      )

  @GetMapping("/{id}")
  fun getOne(@PathVariable id: Long): ResponseEntity<*> =
    getUser.execute(id)
      .fold(
        onFailure = { err ->
          val status = when (err) {
            is HttpClientErrorException.Unauthorized -> HttpStatus.UNAUTHORIZED
            is IllegalArgumentException -> HttpStatus.CONFLICT
            else -> HttpStatus.INTERNAL_SERVER_ERROR
          }
          ResponseEntity.status(status).body(null)
        },
        onSuccess = { usr ->
          ResponseEntity.status(HttpStatus.OK).body(usr.also { usr?.loadRole() })
        }
      )

  @GetMapping
  fun getAll(
    @PageableDefault(size = 10, sort = ["createdAt"], direction = Sort.Direction.DESC) pagination: Pageable
  ): ResponseEntity<Page<User>> =
    getAll.execute(pagination)
      .fold(
        onFailure = { err ->
          val status = when (err) {
            is HttpClientErrorException.Unauthorized -> HttpStatus.UNAUTHORIZED
            is IllegalArgumentException -> HttpStatus.CONFLICT
            else -> HttpStatus.INTERNAL_SERVER_ERROR
          }

          ResponseEntity.status(status).body(null)
        },
        onSuccess = { usr -> ResponseEntity.status(HttpStatus.OK).body(usr) }
      )

}
